types/crypto-js/index.d.ts

```diff
@@ -1,171 +1,1731 @@
+     *
+     * @return The clone.
+     *
+     * @example
+     *
+     *     var clone = bufferedBlockAlgorithm.clone();
+     */
+    clone(): BufferedBlockAlgorithm;
+}
 
-    interface Hasher extends BufferedBlockAlgorithm {
-        update(messageUpdate: WordArray|string): Hasher;
-    }
+/**
+ * Abstract hasher template.
+ */
+interface Hasher {
+    /**
+     * The number of 32-bit words this hasher operates on. Default: 16 (512 bits)
+     */
+    blockSize: number;
+    /**
+     * Resets this hasher to its initial state.
+     *
+     * @example
+     *
+     *     hasher.reset();
+     */
+    reset(): void;
+    /**
+     * Updates this hasher with a message.
+     *
+     * @param messageUpdate The message to append.
+     *
+     * @return This hasher.
+     *
+     * @example
+     *
+     *     hasher.update('message');
+     *     hasher.update(wordArray);
+     */
+    update(messageUpdate: WordArray | string): this;
+    /**
+     * Finalizes the hash computation.
+     * Note that the finalize operation is effectively a destructive, read-once operation.
+     *
+     * @param messageUpdate (Optional) A final message update.
+     *
+     * @return The hash.
+     *
+     * @example
+     *
+     *     var hash = hasher.finalize();
+     *     var hash = hasher.finalize('message');
+     *     var hash = hasher.finalize(wordArray);
+     */
+    finalize(messaegUpdate?: WordArray | string): WordArray;
+}
 
-    interface Cipher extends BufferedBlockAlgorithm {
-        createEncryptor(secretPassphrase: string, option?: CipherOption): Encryptor;
-        createDecryptor(secretPassphrase: string, option?: CipherOption): Decryptor;
-    }
+interface HasherStatic {
+    /**
+     * Initializes a newly created hasher.
+     *
+     * @param cfg (Optional) The configuration options to use for this hash computation.
+     *
+     * @example
+     *
+     *     var hasher = CryptoJS.algo.SHA256.create();
+     */
+    create(cfg?: object): Hasher;
+}
 
-    interface BlockCipher extends Cipher {}
+interface HasherHelper {
+    (message: WordArray | string, cfg?: object): WordArray;
+}
 
-    interface StreamCipher extends Cipher {}
+interface HmacHasherHelper {
+    (message: WordArray | string, key: WordArray | string): WordArray;
+}
 
-    interface CipherHelper {
-        encrypt(message: string | LibWordArray, secretPassphrase: string | WordArray, option?: CipherOption): WordArray;
-        decrypt(encryptedMessage: string | WordArray, secretPassphrase: string | WordArray, option?: CipherOption): DecryptedMessage;
-    }
-    interface Encryptor {
-        process(messagePart: string): string;
-        finalize(): string;
-    }
-    interface Decryptor {
-        process(messagePart: string): string;
-        finalize(): string;
-    }
-    interface LibWordArray {
-        sigBytes: number,
-        words: number[],
-        toString(encoder?: Encoder): string;
-    }
-    export interface WordArray {
```

Comment:
This isn't exported at the top level anymore, and might break things if removed.

Reply:
I don't quite understand. Do you mean `WordArray` should or shouldn't be exported at the top level?

Reply:
Yes, previous versions of these types did that. I'll approve when you remove the v3 folder.